src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -4142,6 +4142,9 @@ namespace Microsoft.FSharp.Core
         [<CompiledName("TypeOf")>]
         let inline typeof<'T> = BasicInlinedOperations.typeof<'T>
 
+        [<CompiledName("NameOf")>]
+        let inline nameof (_: 'T) : string = raise (Exception "may not call directly, should always be optimized away")
```

Comment:
I feel like we could make this have `NoDynamicInvocation"

Reply:
I think it's not needed because the compiler knows of the special rules for this library identifier